temporalio/activity.py

```diff
@@ -0,0 +1,298 @@
+    """
+    return not _current_context.get(None) is None
+
+
+def info() -> Info:
+    """Current activity's info.
+
+    Returns:
+        Info for the currently running activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().info()
+
+
+def heartbeat(*details: Any) -> None:
+    """Send a heartbeat for the current activity.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    heartbeat_fn = _Context.current().heartbeat
+    if not heartbeat_fn:
+        raise RuntimeError("Can only execute heartbeat after interceptor init")
+    heartbeat_fn(*details)
+
+
+def is_cancelled() -> bool:
+    """Whether a cancellation was ever requested on this activity.
+
+    Returns:
+        True if the activity has had a cancellation request, False otherwise.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().cancelled_event.is_set()
+
+
+async def wait_for_cancelled() -> None:
+    """Asynchronously wait for this activity to get a cancellation request.
+
+    Raises:
+        RuntimeError: When not in an async activity.
+    """
+    await _Context.current().cancelled_event.wait()
+
+
+def wait_for_cancelled_sync(timeout: Optional[float] = None) -> None:
+    """Synchronously block while waiting for a cancellation request on this
+    activity.
+
+    This is essentially a wrapper around :py:meth:`threading.Event.wait`.
+
+    Args:
+        timeout: Max amount of time to wait for cancellation.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    _Context.current().cancelled_event.wait_sync(timeout)
+
+
+def is_worker_shutdown() -> bool:
+    """Whether shutdown has been invoked on the worker.
+
+    Returns:
+        True if shutdown has been called on the worker, False otherwise.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    return _Context.current().worker_shutdown_event.is_set()
+
+
+async def wait_for_worker_shutdown() -> None:
+    """Asynchronously wait for shutdown to be called on the worker.
+
+    Raises:
+        RuntimeError: When not in an async activity.
+    """
+    await _Context.current().worker_shutdown_event.wait()
+
+
+def wait_for_worker_shutdown_sync(timeout: Optional[float] = None) -> None:
+    """Synchronously block while waiting for shutdown to be called on the
+    worker.
+
+    This is essentially a wrapper around :py:meth:`threading.Event.wait`.
+
+    Args:
+        timeout: Max amount of time to wait for shutdown to be called on the
+            worker.
+
+    Raises:
+        RuntimeError: When not in an activity.
+    """
+    _Context.current().worker_shutdown_event.wait_sync(timeout)
+
+
+def raise_complete_async() -> NoReturn:
+    """Raise an error that says the activity will be completed
+    asynchronously.
+    """
+    raise _CompleteAsyncError()
+
+
+class _CompleteAsyncError(temporalio.exceptions.TemporalError):
```

Comment:
> Not blocking this PR, do you think we should make this class inherit from BaseException so it's less likely to be caught be application code?

Yes, I will change this since it's a private exception class anyways.

> I saw that asyncio [CancelledError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError) subclasses that

Fun fact, not in Python 3.7 which is why I catch it separately.